torch/lib/THD/base/channels/DataChannelTCP.cpp

```diff
@@ -195,49 +195,60 @@ DataChannelTCP::~DataChannelTCP()
 
 
 void DataChannelTCP::listen(std::uint16_t port = 0) {
-  SYSCHECK(_socket = ::socket(PF_INET, SOCK_STREAM, 0))
+  struct addrinfo hints, *res = NULL;
 
-  struct sockaddr_in addr;
-  socklen_t addr_len = sizeof(addr);
+  memset(&hints, 0x00, sizeof(hints));
+  hints.ai_flags = AI_PASSIVE;
+  hints.ai_family = AF_UNSPEC; // either IPv4 or IPv6
+  hints.ai_socktype = SOCK_STREAM; // TCP
+
+  int err = getaddrinfo(NULL, std::to_string(port).data(), &hints, &res);
+  if (err != 0) {
+    throw std::invalid_argument("cannot find host to listen on");
+  }
 
-  memset(&addr, 0, addr_len);
-  addr.sin_family = AF_INET;
-  addr.sin_port = htons(port);
-  addr.sin_addr.s_addr = INADDR_ANY;
+  // TODO: should we prioritize any `ai_family`?
```

Comment:
Yes. `res` is list with resolved (found) IPv6 and IPv4 addresses, so we can iterate through list check if address has `AF_INET6` family. Then try to make socket. If it fails we can try to find another one IPv6 or try IPv4 addresses if there is no IPv6 and try to make socket once again (until succeeded or out of addresses).